### PR TITLE
Add support for iterable and Iterator

### PR DESCRIPTION
**Pull request by simPod**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Doc updated   | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | #1094
| License       | MIT

- [x] Add `iterable` support
- [x] Add `Iterator` support
- [x] ~Make `iterable` work with jms/meta~

### CODE REVIEW

**Comment by simPod** [comment]
@goetas jms/meta heavily relies on ReflectionClass. However, it's not possible to create reflection class from `iterable`. Any idea how to tackle it?

File: composer.json [hunk 1]
```diff
@@ -52,7 +52,10 @@
     "autoload": {
         "psr-4": {
             "JMS\\Serializer\\": "src/"
-        }
+        },
+        "files": [
+            "src/Functions/iterable_to_array.php"
```
Comment (goetas) [inline composer.json]: Functions can not be auto-loaded, this means that this file will be read on each request. 
You can place this utility function in a static method inside some class, and autoloading will work

File: src/Functions/iterable_to_array.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+use function iterator_to_array;
+
+/**
+ * Copy the iterable into an array. If the iterable is already an array, return it.
+ *
+ * @param bool $use_keys [optional] Whether to use the iterator element keys as index.
+ *
+ * @return array
+ */
+function iterable_to_array(iterable $iterable, bool $use_keys = true) : array
```
Comment (goetas) [inline src/Functions/iterable_to_array.php]: `$use_keys` looks never used in the calling code

**Comment by goetas** [comment]
> @goetas jms/meta heavily relies on ReflectionClass. However, it's not possible to create reflection class from iterable. Any idea how to tackle it?

Can you explain which problems gives this?

**Comment by simPod** [comment]
@goetas ooooooh 👍

File: src/Functions/iterable_to_array.php (hunk 2 above)
Comment (simPod) [reply src/Functions/iterable_to_array.php]: Copied it from library I use so I don't have to include it whole https://github.com/bpolaszek/php-iterable-functions/blob/master/src/iterable-functions.php#L34

I removed `$use_keys`

File: composer.json (hunk 1 above)
Comment (simPod) [reply composer.json]: I tried Functions, WDYT?

Comment (goetas) [reply composer.json]: im good with it

**Comment by goetas** [comment]
Thanks a lot!

**Comment by goetas** [comment]
Well done!

**Pull request merged**